include/seqan3/core/type_traits/lazy.hpp

```diff
@@ -0,0 +1,122 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides lazy template instantiation traits.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <tuple>
+#include <type_traits>
+
+#include <seqan3/core/type_traits/function.hpp>
+
+namespace seqan3
+{
+
+/*!\addtogroup type_traits
+ * \{
+ */
+
+// ----------------------------------------------------------------------------
+// lazy
+// ----------------------------------------------------------------------------
+
+/*!\brief An empty type whose only purpose is to hold an uninstantiated template plus its arguments.
+ * \tparam template_t The uninstantiated template.
+ * \tparam spec_t     The arguments to template_t.
+ */
+template <template <typename ...> typename template_t, typename ...spec_t>
+struct lazy
+{};
+
+// ----------------------------------------------------------------------------
+// instantiate
+// ----------------------------------------------------------------------------
+
+/*!\brief A transformation trait that instantiates seqan3::lazy types. Base template is the identity transformation.
+ * \tparam t The type to operate on.
+ * \implements seqan3::TransformationTrait
+ */
+template <typename t>
+struct instantiate
+{
+    //!\brief Return type of the trait [the identity by default].
+    using type = t;
+};
+
+/*!\brief A transformation trait that instantiates seqan3::lazy types.
+ * \tparam template_t The uninstantiated template.
+ * \tparam spec_t     The arguments to template_t.
+ * \implements seqan3::TransformationTrait
+ */
+template <template <typename ...> typename template_t, typename ...spec_t>
+struct instantiate<lazy<template_t, spec_t...>>
+{
+    //!\brief Return type of the trait [instantiates the template arguments].
+    using type = template_t<spec_t...>;
+};
+
+/*!\brief A transformation trait that instantiates seqan3::lazy types. Transformation trait shortcut.
```

Comment:
```suggestion
/*!\brief A transformation trait that instantiates seqan3::lazy types (TransformationTrait shortcut).
```
We do it like this in the other `core/type_traits/*` headers, but I don't care too much.